syntax = "proto3";

import "options.proto";
import "style.proto";
import "math.proto";

package xviz.v2.primitive;

message Base
{
//    option (xviz_json_schema) = "primitives/base";
    
    string object_id = 1;
    
    repeated string classes = 2;
    
    style.ObjectValue style = 3;
}

message Circle
{
    option (xviz_json_schema) = "primitives/circle";

    Base base = 1;
    
    repeated double center = 2;
    
    double radius_m = 3;
}

message Image
{
    option (xviz_json_schema) = "primitives/image";
    
    Base base = 1;
    
    repeated double position = 2;

    bytes data = 3;
    
    uint32 width_px = 4;

    uint32 height_px = 5;
}

message Point
{
    option (xviz_json_schema) = "primitives/point";

    Base base = 1;
    
    /// Flattened X, Y, Z list
    repeated double points = 2;

    /// Flattened list of (R, G, B) or (R, G, B, A)
    repeated uint32 colors = 3;
}

message Polygon
{
    option (xviz_json_schema) = "primitives/polygon";

    Base base = 1;
    
    /// Flattened X, Y, Z list
    repeated double vertices = 2;
}

message Polyline
{
    option (xviz_json_schema) = "primitives/polyline";

    Base base = 1;

    /// Flattened X, Y, Z list
    repeated double vertices = 2;

    /// Flattened (R, G, B) or (R, G, B, A) list
    repeated uint32 colors = 3;
}

message Stadium
{
    option (xviz_json_schema) = "primitives/stadium";

    Base base = 1;

    repeated double start = 2;

    repeated double end = 3;

    double radius_m = 4;
}

message Text
{
    option (xviz_json_schema) = "primitives/text";

    Base base = 1;

    repeated double position = 2;

    string text = 3;
}
